src/hotspot/share/runtime/objectMonitor.hpp

```diff
@@ -151,7 +141,7 @@ class ObjectMonitor : public CHeapObj<mtObjectMonitor> {
   // ParkEvent of unblocker thread.
   static ParkEvent* _vthread_unparker_ParkEvent;
 
-  // The sync code expects the metadata field to be at offset zero (0).
+  // Because of frequent access, the the metadata field is at offset zero (0).
```

Comment:
```suggestion
  // Because of frequent access, the metadata field is at offset zero (0).
```

Reply:
Isn't it more than that though? Even though we now calculate the offset in the code that uses this field, isn't it still a compile-time constant value of zero and so generates more efficient code compared to adding an offset that happens to be zero?

Reply:
Well the point is that with a known zero offset you just have a load of the base address, you don't have to do `base + offset` either explicitly, or as part of an index load.

Reply:
Don't know.  For example, is this instruction more efficient with a zero offset (aarch64)?
  ldr(r0, Address(c_rarg1, BasicObjectLock::obj_offset()));

The ObjectMonitor::metadata_offset() isn't called by our assembly code in c2_MacroAssembler, it's only used in a gtest.  So if the underlying C++ compiler generates better code for the static offset at zero, then having metadata at zero offset makes sense.  I still wonder if we access the metadata more than owner field though.

In any case, I'm not going to change anything unless I find evidence it would be worth changing.

Reply:
>  For example, is this instruction more efficient with a zero offset (aarch64)?
`ldr(r0, Address(c_rarg1, BasicObjectLock::obj_offset()));`

Apparently not on Aarch64. The Address constructor doesn't look at the actual offset (which is known at code generation time) and still generates a base_plus_offset mode load instruction. x86_64 ndoesn't seem to look at it either but it is hard to track through to see what the actual real cpu instruction will be.

The `metadata` field used to just be the `markWord` field, so given an object we wanted to be able to update the markWord by just writing to the address, not needing to do a dynamic calculation if the offset was non-zero.  Of course we used to write to the markWord a lot more often when we had biased locking.